File: core-utils/src/test/java/io/appmetrica/analytics/push/coreutils/internal/PushServiceFacadeTest.java
package io.appmetrica.analytics.push.coreutils.internal;

import android.content.Context;
import android.os.Bundle;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;

@RunWith(RobolectricTestRunner.class)
public class PushServiceFacadeTest {

    private Context context;
    private PushServiceFacade.CommandServiceWrapper commandServiceWrapper;
    private Bundle pushMessageBundle = new Bundle();

    @Before
    public void setUp() {
        context = RuntimeEnvironment.application;
        commandServiceWrapper = mock(PushServiceFacade.CommandServiceWrapper.class);
        PushServiceFacade.setJobIntentServiceWrapper(commandServiceWrapper);
        pushMessageBundle.putString(CoreConstants.PushMessage.ROOT_ELEMENT, new JSONObject().toString());
    }

    @Test
    public void testInitTokenShouldSendIntentWithExpectedCommand() {
        PushServiceFacade.initToken(context);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().getString(PushServiceFacade.EXTRA_COMMAND))
            .isEqualTo(PushServiceFacade.COMMAND_INIT_PUSH_TOKEN);
    }

    @Test
    public void testRefreshTokenShouldSendIntentWithExpectedCommand() {
        PushServiceFacade.refreshToken(context);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().getString(PushServiceFacade.EXTRA_COMMAND))
            .isEqualTo(PushServiceFacade.COMMAND_UPDATE_TOKEN);
    }

    @Test
    public void testProcessPushShouldSendIntentWithExpectedCommand() {
        String transport = "my_transport";
        PushServiceFacade.processPush(context, pushMessageBundle, transport);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1))
            .startCommand(any(Context.class), arg.capture(), anyBoolean());
        assertThat(arg.getValue().getString(PushServiceFacade.EXTRA_COMMAND))
            .isEqualTo(PushServiceFacade.COMMAND_PROCESS_PUSH);
        assertThat(arg.getValue().getString(CoreConstants.EXTRA_TRANSPORT)).isEqualTo(transport);
    }

    @Test
    public void testInitPushServiceShouldSendIntentWithExpectedCommand() {
        PushServiceFacade.initPushService(context);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().getString(PushServiceFacade.EXTRA_COMMAND))
            .isEqualTo(PushServiceFacade.COMMAND_INIT_PUSH_SERVICE);
    }

    @Test
    public void testInitPushServiceShouldSendReceivedTime() {
        PushServiceFacade.initPushService(context);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testInitTokenShouldSendReceivedTime() {
        PushServiceFacade.initToken(context);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testInitTokenWithForceShouldSendReceivedTime() {
        PushServiceFacade.initToken(context, true);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testRefreshTokenShouldSendReceivedTime() {
        PushServiceFacade.refreshToken(context);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testRefreshTokenWithForceShouldSendReceivedTime() {
        PushServiceFacade.refreshToken(context, true);
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testRefreshTokenWithRefreshTokenInfoShouldSendReceivedTime() {
        PushServiceFacade.refreshToken(context, new RefreshTokenInfo(false));
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1)).startCommand(any(Context.class), arg.capture());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testProcessPushInfoShouldSendReceivedTime() {
        PushServiceFacade.processPush(context, pushMessageBundle, "transport");
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper, times(1))
            .startCommand(any(Context.class), arg.capture(), anyBoolean());
        assertThat(arg.getValue().containsKey(PushServiceFacade.EXTRA_COMMAND_RECEIVED_TIME)).isTrue();
    }

    @Test
    public void testProcessPushInfoShouldNotStartCommandIfNotOwnPush() {
        PushServiceFacade.processPush(context, new Bundle(), "transport");
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verifyNoInteractions(commandServiceWrapper);
    }

    @Test
    public void processPushWithMinProcessingDelay() throws Exception {
        long value = 13;
        JSONObject root = new JSONObject();
        root.put(CoreConstants.PushMessage.PROCESSING_MIN_TIME, value);
        pushMessageBundle.putString(CoreConstants.PushMessage.ROOT_ELEMENT, root.toString());
        PushServiceFacade.processPush(context, pushMessageBundle, "Some transport");
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper).startCommand(any(Context.class), arg.capture(), anyBoolean());
        assertThat(arg.getValue().getLong(CoreConstants.MIN_PROCESSING_DELAY, -1))
            .isEqualTo(value);
    }

    @Test
    public void processPushWithoutMinProcessingDelay() throws Exception {
        PushServiceFacade.processPush(context, pushMessageBundle, "Some transport");
        ArgumentCaptor<Bundle> arg = ArgumentCaptor.forClass(Bundle.class);
        verify(commandServiceWrapper).startCommand(any(Context.class), arg.capture(), anyBoolean());
        assertThat(arg.getValue().getLong(CoreConstants.MIN_PROCESSING_DELAY, -1)).isEqualTo(-1);
    }
}

File: core-utils/src/test/java/io/appmetrica/analytics/push/coreutils/internal/PushServiceFacadeTest.kt
package io.appmetrica.analytics.push.coreutils.internal

import android.content.Context
import android.os.Bundle
import io.appmetrica.analytics.push.coreutils.internal.commands.Commands
import io.appmetrica.analytics.push.coreutils.internal.commands.PushTokenCommandInfo
import io.appmetrica.analytics.push.coreutils.internal.commands.SystemInfoCommandInfo
import io.appmetrica.analytics.push.coreutils.internal.model.ServiceType
import io.appmetrica.analytics.push.testutils.CommonTest
import org.assertj.core.api.Assertions.assertThat
import org.json.JSONObject
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class PushServiceFacadeTest : CommonTest() {

    private val context: Context = mock()
    private val commandServiceWrapper: CommandServiceWrapper = mock()

    @Before
    fun setUp() {
        PushServiceFacade.setJobIntentServiceWrapper(commandServiceWrapper)
    }

    @Test
    fun initPushService() {
        PushServiceFacade.initPushService(context)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(true))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.InitPushService.COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()
    }

    @Test
    fun initToken() {
        val provider = "some provider"
        PushServiceFacade.initToken(context, provider)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(true))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.INIT_PUSH_TOKEN_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isNull()
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun sendToken() {
        val token = "Some token"
        val provider = "some transport"
        PushServiceFacade.sendToken(context, provider, token)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.UPDATE_PUSH_TOKEN_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isEqualTo(token)
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun sendTokenIfTokenIsNull() {
        val provider = "some transport"
        PushServiceFacade.sendToken(context, provider, null)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.UPDATE_PUSH_TOKEN_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isNull()
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun sendTokenOnRefresh() {
        val token = "Some token"
        val provider = "some transport"
        PushServiceFacade.sendTokenOnRefresh(context, provider, token)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.SEND_PUSH_TOKEN_ON_REFRESH_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isEqualTo(token)
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun sendTokenOnRefreshIfTokenIsNull() {
        val provider = "some transport"
        PushServiceFacade.sendTokenOnRefresh(context, provider, null)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.SEND_PUSH_TOKEN_ON_REFRESH_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isNull()
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun sendTokenManually() {
        val token = "Some token"
        val provider = "some transport"
        PushServiceFacade.sendTokenManually(context, provider, token)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.SEND_PUSH_TOKEN_MANUALLY_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isEqualTo(token)
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun sendTokenManuallyIfTokenIsNull() {
        val provider = "some transport"
        PushServiceFacade.sendTokenManually(context, provider, null)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.SendPushToken.SEND_PUSH_TOKEN_MANUALLY_COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val pushTokenCommandInfoBundle = bundle.getBundle(Commands.SendPushToken.EXTRA_INFO)!!
        val pushTokenCommandInfo = PushTokenCommandInfo.fromBundle(pushTokenCommandInfoBundle)

        assertThat(pushTokenCommandInfo.provider).isEqualTo(provider)
        assertThat(pushTokenCommandInfo.token).isNull()
        assertThat(pushTokenCommandInfo.force).isFalse()
    }

    @Test
    fun processPushIfNeedService() {
        val transport = "my_transport"
        val processingMinTime = 13L
        val pushMessageBundle = Bundle().apply {
            putString(
                CoreConstants.PushMessage.ROOT_ELEMENT,
                JSONObject()
                    .put(CoreConstants.PushMessage.PROCESSING_MIN_TIME, processingMinTime)
                    .toString()
            )
        }

        PushServiceFacade.processPush(context, pushMessageBundle, transport)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(true))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.ProcessPush.COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()
        assertThat(bundle.getString(CoreConstants.EXTRA_TRANSPORT))
            .isEqualTo(transport)
        assertThat(bundle.getLong(CoreConstants.MIN_PROCESSING_DELAY))
            .isEqualTo(processingMinTime)
        assertThat(bundle.keySet()).containsAnyElementsOf(pushMessageBundle.keySet())
    }

    @Test
    fun processPushIfNoNeedService() {
        val transport = "my_transport"
        val processingMinTime = 13L
        val pushMessageBundle = Bundle().apply {
            putString(
                CoreConstants.PushMessage.ROOT_ELEMENT,
                JSONObject()
                    .put(CoreConstants.PushMessage.PROCESSING_MIN_TIME, processingMinTime)
                    .put(CoreConstants.PushMessage.SERVICE_TYPE, ServiceType.PROVIDER_SERVICE.value)
                    .toString()
            )
        }

        PushServiceFacade.processPush(context, pushMessageBundle, transport)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(false))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.ProcessPush.COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()
        assertThat(bundle.getString(CoreConstants.EXTRA_TRANSPORT))
            .isEqualTo(transport)
        assertThat(bundle.getLong(CoreConstants.MIN_PROCESSING_DELAY))
            .isEqualTo(processingMinTime)
        assertThat(bundle.keySet()).containsAnyElementsOf(pushMessageBundle.keySet())
    }

    @Test
    fun processPushIfNoProcessingMinTime() {
        val transport = "my_transport"
        val pushMessageBundle = Bundle().apply {
            putString(
                CoreConstants.PushMessage.ROOT_ELEMENT,
                JSONObject().toString()
            )
        }

        PushServiceFacade.processPush(context, pushMessageBundle, transport)
        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(true))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.ProcessPush.COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()
        assertThat(bundle.getString(CoreConstants.EXTRA_TRANSPORT))
            .isEqualTo(transport)
        assertThat(bundle.containsKey(CoreConstants.MIN_PROCESSING_DELAY)).isFalse()
        assertThat(bundle.keySet()).containsAnyElementsOf(pushMessageBundle.keySet())
    }

    @Test
    fun processPushIfNotOwnPush() {
        PushServiceFacade.processPush(context, Bundle(), "transport")

        verifyNoInteractions(commandServiceWrapper)
    }

    @Test
    fun sendSystemInfo() {
        val statusChangeTime = 24523L
        PushServiceFacade.sendSystemInfo(context, statusChangeTime)

        val arg = argumentCaptor<Bundle>()

        verify(commandServiceWrapper).startCommand(eq(context), arg.capture(), eq(true))

        val bundle = arg.firstValue

        assertThat(bundle.getString(Commands.EXTRA_COMMAND))
            .isEqualTo(Commands.UpdateSystemInfo.COMMAND_ACTION)
        assertThat(bundle.containsKey(Commands.EXTRA_COMMAND_RECEIVED_TIME)).isTrue()

        val systemInfoCommandInfoBundle = bundle.getBundle(Commands.UpdateSystemInfo.EXTRA_INFO)!!
        val systemInfoCommandInfo = SystemInfoCommandInfo.fromBundle(systemInfoCommandInfoBundle)

        assertThat(systemInfoCommandInfo.statusChangeTime).isEqualTo(statusChangeTime)
    }
}
